### PR TITLE
Layout 2020: Optimize collapsible margin lookahead

### PR DESCRIPTION
**Pull request by Loirooriol**
Every time that we would lay out a block box that could collapse its top margin with its contents, we would do a lookahead to compute the resulting margin in order to place floats correctly.

The problem is that this lookahead could iterate several descendants, but then when laying these we would run the lookahead again.

This patch restricts the lookahead to boxes that either aren't collapsing their top margin with their parent, or that have 'clear' different than 'none' (since clearance prevents collapsing margins with the parent).

Since the lookahead stops iterating when it finds a box that doesn't collapse its top margin with its parent, or whose 'clear' isn't 'none', this should ensure that lookahead never handles the same box twice.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes do not require tests because there shouldn't be any change in behavior

### CODE REVIEW

**Review by mrobinson — APPROVED** [review]
Looks good. Thank you!

File: components/layout_2020/flow/mod.rs [hunk 1]
```diff
@@ -671,13 +677,19 @@ fn layout_in_flow_non_replaced_block_level(
         None => parent_containing_block_position_info = None,
         Some(ref mut sequential_layout_state) => {
             let mut block_start_margin = CollapsedMargin::new(margin.block_start);
-            if start_margin_can_collapse_with_children {
+            let collapsible_with_parent_start_margin = collapsible_with_parent_start_margin.expect(
+                "We should know whether we are collapsing the block start margin with the parent \
+                when laying out sequentially",
+            ).0 && style.get_box().clear == Clear::None;
```
[the remark this answers is not in the document]
Comment (Loirooriol) [reply components/layout_2020/flow/mod.rs:695]: Added a comment explaining the `clear` check

**Pull request merged**